enhancements/baremetal/continuous-hw-introspection.md

```diff
@@ -0,0 +1,154 @@
+operational status of a server.
+
+### Non-Goals
+
+This proposal does not include any changes to the BareMetalHost CRD.
+
+## Proposal
+
+This enhancement proposes to add a new DaemonSet started by the machine-api-operator
+only for the BareMetal platform. This DaemonSet would run a new container image
+(ironic-hardware-inventory-recorder) on all master and worker nodes. With the new
+addition, it would become possible to have detailed information regarding the
+hardware connected to both master and worker nodes in a metal3 deployment and not
+just when the worker nodes are provisioned. This also allows the user to keep
+track of any changes in hardware capabilities of servers in a metal3 deployment.
+
+Additional details regarding enhancements made to the underlying service required
+for this feature can be found in [3].
+
+### User Stories [optional]
+
+#### Story 1
+
+As an administrator tasked with managing the physical infrastructure, I would like to
+know the current status of the hardware and its capabilities for each of the baremetal
+servers in my Baremetal IPI deployment.
+
+#### Story 2
+
+As an administrator tasked with managing the physical infrastructure, I would like to know
+of any changes in hardware status and capabilities of each baremetal server while it is part
+of my Baremetal IPI deployment.
+
+### Implementation Details/Notes/Constraints [optional]
+
+Continuous hardware introspection involves periodically reading information about baremetal
+server hardware by talking to the server directly. This information includes details
+about devices connected to each server like its CPUs, NICs and storage devices. The
+actual task of hardware introspection is performed by an agent that needs to collect
+information on all master and worker nodes and hence needs to run as a DaemonSet.
+
+This agent then periodically pushes data back to a service running with the metal3 pod. The
+baremetal-operator, also running within the metal3 pod, needs to poll this service for
+hardware introspection data when it is not busy reconciling. The baremetal-operator then
+updates the status on their respective BareMetalHost CRs.
+
+Hardware inspection is curently done on the worker nodes once when it is initially registered
+with the cluster. This information could get outdated unless this introspection is performed at
+regular intervals. Currently, there is no hardware introspection data on the master nodes and
+implemnting this feature will give us access to that data. Details of the underlying service
+used to perform the continuous hardware inspection can be found at [3]. Information on how
+baremetal-operator interacts with this underlying service is detailed in [4].
+
+The machine-api-operator would now be responsible for deploying the DaemonSet that would
+run the service that performs continuous hardware introspection on baremetal hosts.The container
+image for the DaemonSet can be found at [2]. Also, if for some reason, a previously "ready"
```

Comment:
It's not wrong, but I'm not sure why this part about metal3 restarting is included here. Is that a special case we're worried about?

Reply:
No. Just trying to be explicit about expected behavior. I could remove it if the enhancement doc is not the right place for this.